### DISCUSSION TITLE
How would Valhalla be used on a mobile device?

### DISCUSSION THREAD

**Discussion by JonasVautherin** [valhalla/valhalla#4746] [Q&A]
I have read a few [related](https://github.com/valhalla/valhalla/discussions/4509) [conversations](https://github.com/valhalla/valhalla/issues/4704), but it is still a bit unclear to me how Valhalla would be used in a mobile app.

Say I compile Valhalla for Android by disabling everything I can from [here](https://github.com/valhalla/valhalla/blob/master/docs/docs/building.md#build-configuration-all-platforms) and I get a `libvalhalla.so`. Let's say that I will use it from C++ (so no worries about e.g. JNI).

My understanding is that I somehow need to create a `.pbf` file that would represent routing data for parts of the world, so let's also assume I have a `switzerland.pbf` file that I manage to generate.

Then what? Is there a C++ API somewhere in the headers of `libvalhalla.so` that would allow me to feed it `switzerland.pbf` and generate a route? Would it be a one shot call or would Valhalla run as a server? Where would I find this API? Are there examples somewhere?

I suppose an alternative would be to run Valhalla as an HTTP server onboard the mobile device, such that the navigation SDKs could connect to it just like if it was a remote server. Would that be better (I'm mostly thinking about the integration with navigation SDKs here)?

Thanks in advance for the insights :blush:.

**Comment by ianthetechie** [comment]
Hey Jonas :) This is something that [Jacob](https://github.com/Archdoog) is working on: https://github.com/Rallista/valhalla-mobile. The idea is pretty much what you said: integration with nav SDKs (Ferrostar in this case).

> My understanding is that I somehow need to create a .pbf file that would represent routing data for parts of the world, so let's also assume I have a switzerland.pbf file that I manage to generate.
>
> Then what? Is there a C++ API somewhere in the headers of libvalhalla.so that would allow me to feed it switzerland.pbf and generate a route? Would it be a one shot call or would Valhalla run as a server? Where would I find this API? Are there examples somewhere?

Interesting question.... so actually the raw OSM PBF file is quite a few steps away from usable for routing. I think it could be possible to generate tiles on a mobile device, but I'm the wrong person to speculate on that.

Instead, I'll just say how I'd approach it. I would generate tiles periodically server-side and then build a mechanism for shipping them over to the client. What this looks like will be a bit app-specific. Do you want neat extracts for regions? Do you want to just load individual tiles as needed and refresh when you can? This will determine whether you build some sort of "tile API" or ship whole tarballs. In either case, the mobile app wouldn't consume the OSM data and generate tiles; it would just need a valhalla config that points it at the pregenerated tiles. Then you basically call the route function exposed via a minimal wrapper.

Back to the question of why not to generate on-device real quick.... aside from this potentially being extremely resource-intensive, it's also going to make it difficult to get tiles with other data built-in. You'll need an admin database during tile generation, which takes space. Same goes for a timezone database. And if you care about elevation-influenced routing (or elevation profiles), you'll need elevation tiles too (which can be quite large). So while it's probably _possible_ to generate tiles locally, I'd strongly consider generating them on a server.

Here's a quick documentation section on how to get started: https://valhalla.github.io/valhalla/building/#running-valhalla-server-on-unix.

**Comment by JonasVautherin** [comment]
Thanks a lot for the insights! First, let me mention that I don't have a particular app in mind, I'm mostly exploring a bit :blush:. I guess I would be happy to be able to generate a route offline from an app with Ramani-Maps (I am a fan of offline maps, though :grin:).

I had a quick look at valhalla-mobile, which seems to be in agreement with what you said:

> It currently only exposes the route function for the primary purpose of generating turn by turn navigation routes using a downloaded pre-parsed valhalla tileset.

From what you suggest, do I understand correctly that "Valhalla tiles" would be generated offline and downloaded in the app a bit like Organic Maps can download maps for a region? I think that's exactly the kind of UX I was hoping for. Just like in OSMAnd I can download an area, and separately I can download the contour lines for that area, here I could imagine downloading the routing data for that area.

> Here's a quick documentation section on how to get started: https://valhalla.github.io/valhalla/building/#running-valhalla-server-on-unix.

That's the part I don't really get with valhalla-mobile. On a Linux server you run the service with something like:

```sh
valhalla_service valhalla.json 1
```

And then you can query it with something like:

```sh
curl http://localhost:8002/route --data '{"locations":[{"lat":47.365109,"lon":8.546824,"type":"break","city":"Zürich","state":"Altstadt"},{"lat":47.108878,"lon":8.394801,"type":"break","city":"6037 Root","state":"Untere Waldstrasse"}],"costing":"auto","directions_options":{"units":"miles"}}' | jq '.'
```

But valhalla-mobile doesn't seem to run a server. Instead does it instantiate Valhalla for each request, with [this "valhalla-actor"](https://github.com/Rallista/valhalla-mobile/blob/main/src/wrapper/valhalla_actor.cpp#L16)? I guess then a `request` could be the JSON part of the curl request above, i.e.:

```
'{"locations":[{"lat":47.365109,"lon":8.546824,"type":"break","city":"Zürich","state":"Altstadt"},{"lat":47.108878,"lon":8.394801,"type":"break","city":"6037 Root","state":"Untere Waldstrasse"}],"costing":"auto","directions_options":{"units":"miles"}}'
```

And then somehow the "config_path" argument of `route(...)` has to point to some kind of config file that says where the tiles are? :thinking:

I'll have a closer look at Valhalla-Mobile :blush:. Thanks again!

**Comment by SanderKornev** [comment]
Hello @JonasVautherin , @ianthetechie.  
I also wonder if I can create an iOS app with an embedded offline map for the specific small region, let's say a square area of 5 to 5 kilometers. The main feature of the app is offline routing in this specific region.
Is it possible to do with https://github.com/Rallista/valhalla-mobile ?
I'd appreciate any help.

**Comment by HarelM** [comment]
For anyone reading this in the future, I've just finished integrating Rallista's valahalla-mobile into my app (can be found at mapeak.com) with good results. Our app allows downloading pre-built zoom level 7 tiles for offline usage including valhalla tiles of that area and users can now use offline routing.
So definitely possible nowadays.

**Reply by nilsnolde** [reply]
@HarelM you wanna contact me on nilsnolde@proton.me to set up a meeting? I might have a mutually beneficial proposal for you:)

**Reply by nilsnolde** [reply]
hm, didn't know rallista handles the offline tiles. that's kinda what my proposal is around. anyways, would still be good to meet.
